src/ray/core_worker/transport/direct_actor_transport.cc

```diff
@@ -0,0 +1,217 @@
+      gcs_client_(gcs_client),
+      client_call_manager_(io_service),
+      store_provider_(
+          object_interface.CreateStoreProvider(StoreProviderType::LOCAL_PLASMA)) {
+  RAY_CHECK_OK(SubscribeActorUpdates());
+}
+
+Status CoreWorkerDirectActorTaskSubmitter::SubmitTask(
+    const TaskSpecification &task_spec) {
+  if (HasByReferenceArgs(task_spec)) {
+    return Status::Invalid("direct actor call only supports by-value arguments");
+  }
+
+  RAY_CHECK(task_spec.IsActorTask());
+  const auto &actor_id = task_spec.ActorId();
+
+  const auto task_id = task_spec.TaskId();
+  const auto num_returns = task_spec.NumReturns();
+
+  auto request = std::unique_ptr<rpc::PushTaskRequest>(new rpc::PushTaskRequest);
+  request->set_task_id(task_spec.TaskId().Binary());
+  request->mutable_task_spec()->CopyFrom(task_spec.GetMessage());
+
+  std::unique_lock<std::mutex> guard(rpc_clients_mutex_);
+  auto iter = actor_states_.find(actor_id);
+  if (iter == actor_states_.end() || iter->second == ActorTableData::RECONSTRUCTING) {
+    // Actor is not yet created, or is being reconstructing, cache the request
+    // and submit after actor is alive.
+    auto pending_request = std::unique_ptr<PendingTaskRequest>(
+        new PendingTaskRequest(task_id, num_returns, std::move(request)));
+    pending_requests_[actor_id].emplace_back(std::move(pending_request));
+    return Status::OK();
+  } else if (iter->second == ActorTableData::ALIVE) {
+    // Actor is alive, submit the request.
+    // RAY_CHECK(rpc_clients_.count(actor_id) > 0);
+
+    if (rpc_clients_.count(actor_id) == 0) {
+      RAY_CHECK(actor_locations_.count(actor_id) > 0);
+      const auto &actor_location = actor_locations_[actor_id];
+      HandleActorAlive(actor_id, actor_location.first, actor_location.second);
+    }
+
+    // Submit request.
+    auto &client = rpc_clients_[actor_id];
+    return PushTask(*client, *request, task_id, num_returns);
+  } else {
+    // Actor is dead, treat the task as failure.
+    RAY_CHECK(iter->second == ActorTableData::DEAD);
+    TreatTaskAsFailed(task_id, num_returns, rpc::ErrorType::ACTOR_DIED);
+    return Status::IOError("actor is dead or being reconstructed");
+  }
+}
+
+Status CoreWorkerDirectActorTaskSubmitter::SubscribeActorUpdates() {
+  // Register a callback to handle actor notifications.
+  auto actor_notification_callback = [this](const ActorID &actor_id,
+                                            const ActorTableData &actor_data) {
+    if (actor_data.state() == ActorTableData::ALIVE) {
+      RAY_LOG(INFO) << "received notification on actor alive, actor_id: " << actor_id
+                    << ", ip address: " << actor_data.ip_address()
+                    << ", port: " << actor_data.port();
+
+      std::unique_lock<std::mutex> guard(rpc_clients_mutex_);
+      actor_states_[actor_id] = actor_data.state();
+      actor_locations_[actor_id] =
+          std::make_pair(actor_data.ip_address(), actor_data.port());
```

Comment:
Can you combine actor_states_ and actor_locations_ into one map of actor_id -> ActorTableData?

Reply:
Yes that sounds better.